File: python/src/revenant/core/signing.py
"""
Core signing functions — detached CMS and embedded PDF signatures.

All signing functions accept a SigningTransport, making them transport-agnostic.
Use network.SoapSigningTransport(url) to create a transport instance.
"""

from __future__ import annotations

__all__ = [
    "EmbeddedSignatureOptions",
    "sign_data",
    "sign_hash",
    "sign_pdf_detached",
    "sign_pdf_embedded",
]

import logging
from dataclasses import dataclass
from typing import TYPE_CHECKING

from ..constants import DEFAULT_TIMEOUT_SOAP, PDF_MAGIC
from ..errors import PDFError, RevenantError
from .appearance import compute_optimal_height, compute_optimal_width, get_font
from .pdf import (
    SIG_HEIGHT,
    SIG_WIDTH,
    compute_byterange_hash,
    insert_cms,
    prepare_pdf_with_sig_field,
    verify_embedded_signature,
)

if TYPE_CHECKING:
    from ..network.protocol import SigningTransport

_logger = logging.getLogger(__name__)


@dataclass(frozen=True)
class EmbeddedSignatureOptions:
    """Options for embedded PDF signature placement and appearance.

    Bundles all appearance and positioning parameters for
    :func:`sign_pdf_embedded` into a single reusable object.

    Attributes:
        page: Page for the signature -- 0-based int, "first", or "last".
        position: Preset position name ("bottom-right", "br", etc.).
            Ignored when x/y are provided explicitly.
        x: Manual x-coordinate (PDF points, origin = bottom-left).
        y: Manual y-coordinate (PDF points, origin = bottom-left).
        w: Signature field width in PDF points.
        h: Signature field height in PDF points.
        reason: Signature reason string.
        name: Signer display name (used for PDF /Name entry).
        image_path: Optional path to a PNG/JPEG signature image.
        fields: Ordered display strings for the signature appearance.
        visible: If False, create an invisible signature.
        font: Font registry key (e.g. "noto-sans", "ghea-grapalat").
    """

    page: int | str = "last"
    position: str = "bottom-right"
    x: float | None = None
    y: float | None = None
    w: float = SIG_WIDTH
    h: float = SIG_HEIGHT
    reason: str = "Signed with Revenant"
    name: str | None = None
    image_path: str | None = None
    fields: list[str] | None = None
    visible: bool = True
    font: str | None = None


_OPTIONS_FIELDS = frozenset(
    (
        "page",
        "position",
        "x",
        "y",
        "w",
        "h",
        "reason",
        "name",
        "image_path",
        "fields",
        "visible",
        "font",
    )
)


def _resolve_options(
    options: EmbeddedSignatureOptions | None,
    kwargs: dict[str, object],
) -> EmbeddedSignatureOptions:
    """Merge explicit keyword arguments into an options instance.

    Keyword arguments override the corresponding fields in *options*.
    Unknown keys raise TypeError.
    """
    unknown = set(kwargs) - _OPTIONS_FIELDS
    if unknown:
        raise TypeError(f"Unexpected keyword arguments: {', '.join(sorted(unknown))}")

    if options is None:
        options = EmbeddedSignatureOptions()

    if not kwargs:
        return options

    from dataclasses import replace

    overrides = {k: v for k, v in kwargs.items() if k in _OPTIONS_FIELDS}
    return replace(options, **overrides)


def _validate_pdf(pdf_bytes: bytes) -> None:
    """Raise PDFError if bytes don't look like a PDF."""
    if not pdf_bytes or not pdf_bytes.startswith(PDF_MAGIC):
        raise PDFError("Input does not appear to be a PDF file.")


def sign_pdf_detached(
    pdf_bytes: bytes,
    transport: SigningTransport,
    username: str,
    password: str,
    timeout: int = DEFAULT_TIMEOUT_SOAP,
) -> bytes:
    """
    Sign a PDF document — returns a detached CMS/PKCS#7 signature.

    The transport handles the actual communication with the signing service.

    Args:
        pdf_bytes: Raw PDF file content.
        transport: SigningTransport implementation.
        username, password: Revenant credentials.
        timeout: Request timeout in seconds.

    Returns:
        Detached CMS/PKCS#7 signature (DER-encoded).
    """
    _validate_pdf(pdf_bytes)
    return transport.sign_pdf_detached(pdf_bytes, username, password, timeout)


def sign_hash(
    hash_bytes: bytes,
    transport: SigningTransport,
    username: str,
    password: str,
    timeout: int = DEFAULT_TIMEOUT_SOAP,
) -> bytes:
    """
    Sign a pre-computed hash.

    Args:
        hash_bytes: Pre-computed hash (typically 20-byte SHA-1).
        transport: SigningTransport implementation.
        username, password: Revenant credentials.
        timeout: Request timeout in seconds.

    Returns:
        CMS/PKCS#7 signature (DER-encoded) over the provided hash.
    """
    from ..constants import SHA1_DIGEST_SIZE

    if len(hash_bytes) != SHA1_DIGEST_SIZE:
        raise RevenantError(
            f"Expected {SHA1_DIGEST_SIZE}-byte SHA-1 hash, got {len(hash_bytes)} bytes."
        )
    return transport.sign_hash(hash_bytes, username, password, timeout)


def sign_data(
    data_bytes: bytes,
    transport: SigningTransport,
    username: str,
    password: str,
    timeout: int = DEFAULT_TIMEOUT_SOAP,
) -> bytes:
    """
    Sign arbitrary data.

    The transport's server computes SHA-1 internally and returns a CMS/PKCS#7
    signature with correct DigestAlgorithm and messageDigest attributes.

    Args:
        data_bytes: Raw data to sign (any size).
        transport: SigningTransport implementation.
        username, password: Revenant credentials.
        timeout: Request timeout in seconds.

    Returns:
        CMS/PKCS#7 signature (DER-encoded) over the data.
    """
    if not data_bytes:
        raise RevenantError("Cannot sign empty data.")
    return transport.sign_data(data_bytes, username, password, timeout)


def sign_pdf_embedded(
    pdf_bytes: bytes,
    transport: SigningTransport,
    username: str,
    password: str,
    timeout: int = DEFAULT_TIMEOUT_SOAP,
    options: EmbeddedSignatureOptions | None = None,
    **kwargs: object,
) -> bytes:
    """
    Sign a PDF with an embedded signature.

    Uses the data-then-sign workflow:
    1. Prepare PDF with empty signature field
    2. Extract ByteRange data (everything except hex placeholder)
    3. Send ByteRange data to transport for signing
    4. Insert CMS into the PDF
    5. Verify the signature

    Requires pikepdf.

    Args:
        pdf_bytes: Raw PDF file content.
        transport: SigningTransport implementation.
        username, password: Revenant credentials.
        timeout: Request timeout in seconds.
        options: Signature placement and appearance options.
            If None, uses defaults from EmbeddedSignatureOptions.
            Individual keyword arguments (page, position, x, y, w, h,
            reason, name, image_path, fields, visible, font) are also
            accepted and override the corresponding options fields.

    Returns:
        Complete PDF with embedded signature.
    """
    _validate_pdf(pdf_bytes)

    opts = _resolve_options(options, kwargs)

    w = opts.w
    h = opts.h
    if w <= 0 or h <= 0:
        raise PDFError(f"Signature dimensions must be positive, got w={w}, h={h}")
    if opts.x is not None and opts.x < 0:
        raise PDFError(f"Signature x-coordinate must be non-negative, got {opts.x}")
    if opts.y is not None and opts.y < 0:
        raise PDFError(f"Signature y-coordinate must be non-negative, got {opts.y}")

    _logger.info(
        "Signing PDF (embedded, %s): %d bytes, page=%s, position=%s",
        "visible" if opts.visible else "invisible",
        len(pdf_bytes),
        opts.page,
        opts.position,
    )

    # Auto-size dimensions if fields are available (visible mode only)
    if opts.visible:
        font_obj = get_font(opts.font)
        has_img = opts.image_path is not None
        if opts.fields:
            if w == SIG_WIDTH:
                w = compute_optimal_width(opts.fields, h, has_image=has_img, font=font_obj)
                _logger.debug("Adaptive signature width: %.1f pt", w)
            if h == SIG_HEIGHT:
                h = compute_optimal_height(opts.fields, w, has_image=has_img, font=font_obj)
                _logger.debug("Adaptive signature height: %.1f pt", h)

    # Step 1: Prepare PDF with signature field
    _logger.debug("Step 1: Preparing PDF with signature field")
    prepared_pdf, hex_start, hex_len = prepare_pdf_with_sig_field(
        pdf_bytes,
        page=opts.page,
        x=opts.x,
        y=opts.y,
        w=w,
        h=h,
        position=opts.position,
        reason=opts.reason,
        name=opts.name,
        image_path=opts.image_path,
        fields=opts.fields,
        visible=opts.visible,
        font=opts.font,
    )
    _logger.debug("Prepared PDF: %d bytes, hex_start=%d", len(prepared_pdf), hex_start)

    # Step 2: Extract ByteRange data (everything except hex placeholder)
    _logger.debug("Step 2: Extracting ByteRange data")
    before = prepared_pdf[:hex_start]
    after = prepared_pdf[hex_start + hex_len + 1 :]  # +1 for '>'
    br_data = before + after
    _logger.debug("ByteRange data: %d bytes", len(br_data))

    # Step 3: Send ByteRange data to transport for signing
    _logger.debug("Step 3: Sending ByteRange data to transport")
    cms_der = sign_data(br_data, transport, username, password, timeout)
    _logger.debug("Received CMS: %d bytes", len(cms_der))

    # Step 4: Insert CMS into PDF
    _logger.debug("Step 4: Inserting CMS into PDF")
    signed_pdf = insert_cms(prepared_pdf, hex_start, hex_len, cms_der)
    if len(signed_pdf) != len(prepared_pdf):
        raise PDFError(f"insert_cms changed PDF size: {len(prepared_pdf)} -> {len(signed_pdf)}")

    # Step 5: Verify the result
    _logger.debug("Step 5: Verifying signature")
    br_hash = compute_byterange_hash(prepared_pdf, hex_start, hex_len)
    result = verify_embedded_signature(signed_pdf, expected_hash=br_hash)
    if not result["valid"]:
        detail_str = "\n  ".join(result["details"])
        _logger.error("Post-sign verification failed: %s", detail_str)
        raise PDFError(
            f"Post-sign verification FAILED:\n  {detail_str}\n"
            "The signed PDF may be corrupt — not saved."
        )
    _logger.debug("Signature verified successfully")

    _logger.info("Signed PDF complete: %d bytes", len(signed_pdf))
    return signed_pdf

File: typescript/src/core/signing.ts
// SPDX-License-Identifier: Apache-2.0
/**
 * Core signing functions -- detached CMS and embedded PDF signatures.
 *
 * All signing functions accept a SigningTransport, making them
 * transport-agnostic. Use SoapSigningTransport to create an instance.
 */

import { DEFAULT_TIMEOUT_SOAP, PDF_MAGIC, SHA1_DIGEST_SIZE } from "../constants.js";
import { PDFError, RevenantError } from "../errors.js";
import type { SigningTransport } from "../network/protocol.js";
import { computeOptimalHeight, computeOptimalWidth, getFont } from "./appearance/index.js";
import {
  computeByterangeHash,
  insertCms,
  type PrepareOptions,
  preparePdfWithSigField,
  SIG_HEIGHT,
  SIG_WIDTH,
  verifyEmbeddedSignature,
} from "./pdf/index.js";

// -- Helpers ------------------------------------------------------------------

function validatePdf(pdfBytes: Uint8Array): void {
  if (pdfBytes.length < PDF_MAGIC.length) {
    throw new PDFError("Input does not appear to be a PDF file.");
  }
  for (let i = 0; i < PDF_MAGIC.length; i++) {
    if (pdfBytes[i] !== PDF_MAGIC[i]) {
      throw new PDFError("Input does not appear to be a PDF file.");
    }
  }
}

// -- Detached signing ---------------------------------------------------------

/**
 * Sign a PDF document -- returns a detached CMS/PKCS#7 signature.
 */
export async function signPdfDetached(
  pdfBytes: Uint8Array,
  transport: SigningTransport,
  username: string,
  password: string,
  timeout: number = DEFAULT_TIMEOUT_SOAP,
): Promise<Uint8Array> {
  validatePdf(pdfBytes);
  return transport.signPdfDetached(pdfBytes, username, password, timeout);
}

// -- Hash signing -------------------------------------------------------------

/**
 * Sign a pre-computed hash (typically 20-byte SHA-1).
 */
export async function signHash(
  hashBytes: Uint8Array,
  transport: SigningTransport,
  username: string,
  password: string,
  timeout: number = DEFAULT_TIMEOUT_SOAP,
): Promise<Uint8Array> {
  if (hashBytes.length !== SHA1_DIGEST_SIZE) {
    throw new RevenantError(
      `Expected ${SHA1_DIGEST_SIZE}-byte SHA-1 hash, got ${hashBytes.length} bytes.`,
    );
  }
  return transport.signHash(hashBytes, username, password, timeout);
}

// -- Data signing -------------------------------------------------------------

/**
 * Sign arbitrary data. The server computes SHA-1 internally.
 */
export async function signData(
  dataBytes: Uint8Array,
  transport: SigningTransport,
  username: string,
  password: string,
  timeout: number = DEFAULT_TIMEOUT_SOAP,
): Promise<Uint8Array> {
  if (dataBytes.length === 0) {
    throw new RevenantError("Cannot sign empty data.");
  }
  return transport.signData(dataBytes, username, password, timeout);
}

// -- Embedded PDF signing -----------------------------------------------------

/**
 * Sign a PDF with an embedded signature.
 *
 * Uses the data-then-sign workflow:
 * 1. Prepare PDF with empty signature field
 * 2. Extract ByteRange data (everything except hex placeholder)
 * 3. Send ByteRange data to transport for signing
 * 4. Insert CMS into the PDF
 * 5. Verify the signature
 */
export async function signPdfEmbedded(
  pdfBytes: Uint8Array,
  transport: SigningTransport,
  username: string,
  password: string,
  timeout: number = DEFAULT_TIMEOUT_SOAP,
  options: PrepareOptions = {},
): Promise<Uint8Array> {
  validatePdf(pdfBytes);

  let w = options.width ?? SIG_WIDTH;
  let h = options.height ?? SIG_HEIGHT;
  const visible = options.visible ?? true;

  if (w <= 0 || h <= 0) {
    throw new PDFError(`Signature dimensions must be positive, got w=${w}, h=${h}`);
  }
  if (options.x != null && options.x < 0) {
    throw new PDFError(`Signature x-coordinate must be non-negative, got ${options.x}`);
  }
  if (options.y != null && options.y < 0) {
    throw new PDFError(`Signature y-coordinate must be non-negative, got ${options.y}`);
  }

  // Auto-size dimensions if fields are available (visible mode only)
  if (visible && options.fields) {
    const fontObj = await getFont(options.font);
    const hasImg = options.imagePath != null;
    if (w === SIG_WIDTH) {
      w = computeOptimalWidth(options.fields, h, hasImg, fontObj);
    }
    if (h === SIG_HEIGHT) {
      h = computeOptimalHeight(options.fields, w, hasImg, fontObj);
    }
  }

  // Step 1: Prepare PDF with signature field
  const prepareOpts: PrepareOptions = {
    ...options,
    width: w,
    height: h,
  };
  const {
    pdf: preparedPdf,
    hexStart,
    hexLen,
  } = await preparePdfWithSigField(pdfBytes, prepareOpts);

  // Step 2: Extract ByteRange data (everything except hex placeholder)
  const before = preparedPdf.slice(0, hexStart);
  const after = preparedPdf.slice(hexStart + hexLen + 1); // +1 for '>'
  const brData = new Uint8Array(before.length + after.length);
  brData.set(before, 0);
  brData.set(after, before.length);

  // Step 3: Send ByteRange data to transport for signing
  const cmsDer = await signData(brData, transport, username, password, timeout);

  // Step 4: Insert CMS into PDF
  const signedPdf = insertCms(preparedPdf, hexStart, hexLen, cmsDer);
  if (signedPdf.length !== preparedPdf.length) {
    throw new PDFError(`insertCms changed PDF size: ${preparedPdf.length} -> ${signedPdf.length}`);
  }

  // Step 5: Verify the result
  const brHash = computeByterangeHash(preparedPdf, hexStart, hexLen);
  const result = await verifyEmbeddedSignature(signedPdf, brHash);
  if (!result.valid) {
    const detailStr = result.details.join("\n  ");
    throw new PDFError(
      `Post-sign verification FAILED:\n  ${detailStr}\n` +
        "The signed PDF may be corrupt -- not saved.",
    );
  }

  return signedPdf;
}
